File: backend/app/domain/auth/roles.py
"""Los 4 roles funcionales del sistema y su politica de MFA (PURO).

c-76: "proctor" y "revisor" fueron ELIMINADOS del dominio (absorbidos por
COORDINADOR/TUTOR — ver comentarios inline debajo). c-76-2: "admin_examenes"
y "auditor" fueron ELIMINADOS del dominio (absorbidos por ADMIN_SISTEMA — el
dueño del producto decidio que solo debe existir un rol "Admin"). El conteo
de roles vivos bajo de 8 a 6 (c-76) y de 6 a 4 (c-76-2).

Fuente: `03` §RBAC. Codifica como dato de dominio (no como string suelto):
- el conjunto canonico de roles validos,
- cuales exigen MFA obligatorio (acceso a evidencia / administracion, `03`/`08`),
- el rol "estudiante", unico sin MFA (solo accede a su propia sesion / sus datos).

Sin framework ni infraestructura (D1). Estos son enums/constantes de negocio.
"""

from __future__ import annotations

import enum


class Rol(str, enum.Enum):
    """Rol funcional del sistema (`03` §RBAC). Hereda de ``str`` para serializar
    directo en claims/JSON sin perder el control de los valores validos."""

    ESTUDIANTE = "estudiante"
    # (Rol "proctor" ELIMINADO en c-76: el COORDINADOR absorbe la supervision
    # global en vivo + el veredicto. Los usuarios con rol "proctor" fueron
    # remapeados a "coordinador" por la migracion 0068. Un claim de token con
    # "proctor" ya no mapea a ningun Rol de dominio: se descarta en silencio,
    # igual que cualquier rol desconocido — ver parse_rol y Q1 del design c-76.)
    # (Rol "revisor" ELIMINADO en c-76: el TUTOR ya supervisa/observa en vivo
    # dentro de su comision y el COORDINADOR absorbe el veredicto terminal
    # (`revisar_sesion`). Los usuarios con rol "revisor" fueron remapeados a
    # "coordinador" por la migracion 0071. Un claim de token con "revisor" ya
    # no mapea a ningun Rol de dominio: se descarta en silencio, igual que
    # cualquier rol desconocido — mismo precedente que "proctor".)
    COORDINADOR = "coordinador"
    # (Rol "admin_examenes" ELIMINADO en c-76-2: la gestion academica que
    # tenia — examenes/materias/comisiones sin poder de supervision — pasa a
    # ser exclusiva de ADMIN_SISTEMA. Los usuarios con rol "admin_examenes"
    # fueron remapeados a "admin_sistema" por la migracion 0074. Un claim de
    # token con "admin_examenes" ya no mapea a ningun Rol de dominio: se
    # descarta en silencio, mismo precedente que "proctor"/"revisor".)
    ADMIN_SISTEMA = "admin_sistema"
    # (Rol "auditor" ELIMINADO en c-76-2: la capacidad ver_auditoria (solo
    # lectura del registro de auditoria) queda exclusiva de ADMIN_SISTEMA —
    # nunca hubo un endpoint real conectado a la capacidad para "auditor"
    # (require_roles(ADMIN_SISTEMA) hardcodeado en audit_router.py), asi que
    # esta eliminacion no le saca acceso real a nadie. Los usuarios con rol
    # "auditor" fueron remapeados a "admin_sistema" por la migracion 0075.
    # Un claim de token con "auditor" ya no mapea a ningun Rol de dominio: se
    # descarta en silencio, mismo precedente que "proctor"/"revisor".)
    # Gestion academica SIN poder de supervision: carga y administra examenes,
    # materias y comisiones de lo suyo. NO revisa sesiones, NO resuelve casos, NO
    # toca la configuracion del sistema ni la auditoria. Es el rol de quien dicta
    # la materia, no el de quien juzga la integridad de la rendicion — mantener
    # esa separacion es lo que evita que quien pone la nota decida el fraude.
    # (Rol "tutor" — antes llamado "docente"; renombrado en migracion 0060.)
    TUTOR = "tutor"
    # c-78 (E-04, D11): el hueco entre TUTOR (demasiado poco: no crea examenes ni
    # toca el banco) y COORDINADOR (demasiado: emite el VEREDICTO de integridad).
    #
    # El PROFESOR se define por lo que NO puede: emitir veredicto. Esa decision
    # queda EXCLUSIVA del COORDINADOR. Es la misma separacion que ya justifica al
    # TUTOR — quien pone la nota no decide si hubo fraude (regla dura #5) — y sin
    # esa linea PROFESOR y COORDINADOR serian el mismo rol con dos nombres.
    #
    # Puede: crear examenes, gestionar el banco de preguntas, ver estadisticas,
    # ver el registro de sesiones y supervisar en vivo (ver CAPABILITY_ROLES).
    PROFESOR = "profesor"


# Roles que EXIGEN MFA: todo el que accede a evidencia o administracion (`03`,
# `08` §Seguridad). El estudiante es el unico exento (solo su propia sesion/datos).
ROLES_CON_MFA: frozenset[Rol] = frozenset(
    {
        Rol.COORDINADOR,
        Rol.ADMIN_SISTEMA,
        # El tutor administra contenido academico (examenes con sus preguntas y
        # respuestas correctas): es acceso de administracion, exige MFA.
        Rol.TUTOR,
        # c-78: el profesor CREA ese contenido (banco de preguntas incluido) y
        # supervisa en vivo. Con mas acceso que el tutor, MFA no es opcional.
        Rol.PROFESOR,
    }
)

# Roles administrativos de examen (CRUD de Examen, usado por C-07). El tutor
# entra aca: cargar y configurar los examenes de su materia es exactamente su
# trabajo. Lo que NO gana por estar aca es supervision ni configuracion del
# sistema — eso vive en CAPABILITY_ROLES, no en esta lista.
ROLES_ADMIN_EXAMEN: frozenset[Rol] = frozenset(
    {Rol.ADMIN_SISTEMA, Rol.TUTOR, Rol.PROFESOR}
)


def parse_rol(valor: str) -> Rol | None:
    """Convierte un string de claim a ``Rol``, o ``None`` si no es un rol valido.

    Tolerante a roles desconocidos en el token (p. ej. roles internos de Keycloak):
    se descartan sin romper, no se mapean a un rol del dominio."""
    try:
        return Rol(valor)
    except ValueError:
        return None


def rol_exige_mfa(rol: Rol) -> bool:
    """``True`` si el rol exige segundo factor para operar (`03`/`08`)."""
    return rol in ROLES_CON_MFA

File: frontend/src/lib/constants/roles.ts
import type { Rol } from "../types";

// Etiquetas de todos los roles reconocidos — incluye roles legacy para mostrar
// correctamente en badges y tablas si ya existen en la DB.
export const ROL_LABELS: Record<string, string> = {
  estudiante: 'Estudiante',
  tutor: 'Tutor',
  profesor: 'Profesor',
  coordinador: 'Coordinador',
  admin_sistema: 'Admin',
  // Legacy: cuentas viejas que todavía tuvieran el valor "docente" (pre-migración 0060).
  docente: 'Tutor',
};

// Qué puede hacer cada rol, en castellano llano. Se muestra en el formulario de alta.
export const ROL_DESCRIPCIONES: Record<string, string> = {
  estudiante: 'Rinde exámenes. Solo ve lo suyo.',
  tutor: 'Acompaña sus comisiones: cierra notas y supervisa en vivo. NO crea exámenes, NO toca el banco de preguntas y NO administra materias, comisiones ni inscripciones.',
  profesor: 'Administra sus materias: crea comisiones, inscribe alumnos, arma los exámenes y el banco de preguntas, y supervisa en vivo. NO decide si hubo fraude: eso es del coordinador.',
  coordinador: 'Todo lo del profesor más el veredicto de integridad: es el único que puede anular una nota por fraude.',
  admin_sistema: 'Acceso total, incluida la configuración del sistema, los usuarios y la auditoría.',
};

// Roles que puede asignar el formulario de creación/edición de usuarios.
// c-76: 'proctor' y 'revisor' eliminados del dominio — el coordinador absorbe la supervisión global.
// c-76-2: 'admin_examenes' y 'auditor' eliminados del dominio — solo existe un rol "Admin".
export const ROLES_FORMULARIO: Rol[] = ['estudiante', 'tutor', 'profesor', 'coordinador', 'admin_sistema'];

// Todos los roles funcionales reconocidos por el dominio (backend/app/domain/auth/roles.py),
// para poblar selects de FILTRO (a diferencia de ROLES_FORMULARIO, que es solo lo asignable).
export const ROLES_TODOS: Rol[] = ['estudiante', 'tutor', 'profesor', 'coordinador', 'admin_sistema'];

// Valores válidos para type guard en los adapters de auth.
export const ROLES_VALIDOS = Object.keys(ROL_LABELS) as Rol[];

/**
 * Retorna la etiqueta legible del rol. Si la clave no existe, retorna
 * el identificador sin transformar (fallback seguro).
 */
export function getRolLabel(rol: string): string {
  return ROL_LABELS[rol] ?? rol;
}
